corehq/apps/users/user_data.py

```diff
@@ -0,0 +1,136 @@
+from django.utils.functional import cached_property
+
+from corehq.apps.custom_data_fields.models import (
+    COMMCARE_PROJECT,
+    PROFILE_SLUG,
+    CustomDataFieldsDefinition,
+    CustomDataFieldsProfile,
+    is_system_key,
+)
+from corehq.apps.users.views.mobile.custom_data_fields import (
+    CUSTOM_USER_DATA_FIELD_TYPE,
+)
+
+
+class UserDataError(Exception):
+    ...
+
+
+class UserData:
+    def __init__(self, raw_user_data, domain):
+        self._local_to_user = raw_user_data
+        self.domain = domain
+
+    @property
+    def profile_id(self):
+        return self._local_to_user.get(PROFILE_SLUG)
+
+    @property
+    def _provided_by_system(self):
+        return {
+            **(self.profile.fields if self.profile else {}),
+            COMMCARE_PROJECT: self.domain,
+        }
+
+    def to_dict(self):
```

Comment:
I find this function name a bit confusing. It does not convert the UserData to a different format it just combines the values that are held separately internally. I cannot think of a better one of the top of my head though.

Reply:
My thought process there was that `UserData` is meant to be interacted with like a dictionary, though it's the combination of a few different sources.  That is, you can generally do
```python
user_data = UserData(...)
user_data["name"] = "Ethan"
print(user_data["commcare_location_id"])
```
and so forth, but we do need guards around what can and can't be set directly, since some fields care controlled by the profile.  I added the `to_dict()` method for places that just need a normal python dictionary, similar to how `JsonObject` and `Document` subclasses have `to_json()` methods.

Reply:
I suppose `to_dict` is fine